### DISCUSSION TITLE
Is possible to prefill the React Context API in the getInitialProps step?

### DISCUSSION THREAD

**Discussion by thomaschaaf** [vercel/next.js#10639] [Help]
For SSR I need a way to initialize the React Context API.

The Redux Example contains a getInitalProps function: https://github.com/zeit/next.js/blob/canary/examples/with-redux/pages/index.js

The Context Api example doesn't: https://github.com/zeit/next.js/blob/canary/examples/with-context-api/pages/index.js

Is it possible at all or should I migrate to Redux for state management?

**Comment by rafaelalmeidatk** [comment]
You have to get the data on `getInitialProps` and create the provider on your page with this value. Something like this:

```js
const YourPage = ({ foo }) => {
  return (
    <FooProvider foo={foo}>
      {/* Content of the page */}
    </FooProvider>
  )
}

YourPage.getInitialProps = () => {
  const res = await fetch(...);
  const foo = res.json();


  return { foo };
}
```

**Comment by thomaschaaf** [comment]
Thank you for your suggestion but then I am not using the reducer and actions. Do you think those can be used?

**Comment by rafaelalmeidatk** [comment]
Ah I see what you mean. You want to call `dispatch` on `getInitialProps` to send the data to the provider that is above the page, similar to the redux example, right?

I have been thinking about this and I am not sure you could access the `dispatch` of `useReducer` in `getInitialProps` since it is a hook. Using my last snippet as a base, you could still return `foo` from `YourPage.getInitialProps`, but on `_app.js` you can grab this value and automatically add it to the provider, making it available to the page.

**Comment by rafaelalmeidatk** [comment]
Here is an example, I haven't tested it but that's the idea:

```js
// ------------------------------
// pages/mypage.js

const MyPage = () => {
  // useFoo may be useContext(FooContext) too, but I prefer to use custom hooks
  // to make the code cleaner
  const foo = useFoo()

  return <div>Foo: {foo}</div>
}

MyPage.getInitialProps = () => {
  const res = await fetch(...)
  const foo = res.json()

  // This will go to _app, which will add to the provider and make it available
  // to all the components of the page
  return { foo }
}

// ------------------------------
// pages/_app.js

const MyApp = ({ Component, pageProps }) => {
  const { foo, ...actualPageProps } = pageProps

  return (
    // Now foo will be available to all pages
    <FooProvider value={foo}>
      <Component {...actualPageProps} />
    </FooProvider>
  )
}


```

**Reply by devinhalladay** [reply]
Thanks so much for this. You just taught me something that unblocks side projects I've been stuck on for literally two years. I had no idea this was possible.

**Comment by thomaschaaf** [comment]
Yes exactly. Ich have been looking at the unit tests of react to see if there is a way to dispatch it as the functionality is only available as hooks.

**Comment by HaNdTriX** [comment]
Actually the [with-redux](https://github.com/zeit/next.js/blob/canary/examples/with-redux/pages/index.js) optionally prefills the context from `getInitialProps`:

All of this logic is abstracted away in a Higher Order Component:

https://github.com/zeit/next.js/blob/008a558d033328f5dbe6c6557cad686d375945d1/examples/with-redux/lib/redux.js#L33-L54

https://github.com/zeit/next.js/blob/008a558d033328f5dbe6c6557cad686d375945d1/examples/with-redux/lib/redux.js#L7-L14

Please note that we disabled this behaviour on `_app.js` level by propose. Nevertheless if you want to reintroduce this behaviour (initialize hoc on `_app.js` take a look at the `with-apollo` implementation on how this could work.)
